crates/dbsp/src/dynamic/downcast.rs

```diff
@@ -0,0 +1,64 @@
+use std::any::{Any, TypeId};
+
+/// Trait for types that return a reference to self as `dyn Any`.
+///
+/// Used to construct [`DowncastTrait`].
+pub trait AsAny: 'static {
+    fn as_any(&self) -> &dyn Any;
+}
+
+impl<T: 'static> AsAny for T {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
+
+/// A trait for trait objects that can be downcast to a reference
+/// to the concrete type they wrap.
+// NOTE: We don't auto-derive this trait for all types that satisfy
+// the bound.  Instead it needs to be derived manually for trait
+// object types only.
+pub trait DowncastTrait: AsAny {
+    /// Cast trait object reference to a reference to a concrete type `T`.
+    ///
+    /// Panics if the inner value of `self` does not have type `T`.
+    #[inline]
+    fn downcast_checked<T: AsAny>(&self) -> &T {
+        assert_eq!(self.as_any().type_id(), TypeId::of::<T>());
+        unsafe { &*(self as *const _ as *const T) }
+    }
+
+    /// Cast trait object reference to a mutable reference to a concrete type `T`.
+    ///
+    /// Panics if the inner value of `self` does not have type `T`.
+    #[inline]
+    fn downcast_mut_checked<T: AsAny>(&mut self) -> &mut T {
+        //println!("T: {}", std::any::type_name::<T>());
+        assert_eq!((self as &Self).as_any().type_id(), TypeId::of::<T>());
```

Comment:
I think it's a bug in the Rust type checker but it insists on converting `self` into an immutable reference explicitly for this line to compile, so that's what this expression does.